tools/whipinto/src/rtspclient.rs

```diff
@@ -0,0 +1,417 @@
+use anyhow::{anyhow, Result};
+use cli::Codec;
+use md5::{Digest, Md5};
+use portpicker::pick_unused_port;
+use tokio::{
+    io::{AsyncReadExt, AsyncWriteExt},
+    net::TcpStream,
+    time::{sleep, Duration},
+};
+use tracing::info;
+use url::Url;
+
+const USER_AGENT: &str = "whipinto";
+const DEFAULT_RTSP_PORT: u16 = 554;
+
+struct RtspSession {
+    stream: TcpStream,
+    rtsp_url: String,
+    username: String,
+    password: String,
+    realm: String,
+    nonce: String,
+    cseq: u32,
+}
+
+impl RtspSession {
+    async fn new(rtsp_url: &str) -> Result<Self> {
+        let url = Url::parse(rtsp_url)?;
+
+        let username = url.username().to_string();
+        let password = url.password().unwrap_or("").to_string();
+        let host = url
+            .host_str()
+            .ok_or_else(|| anyhow::anyhow!("Host not found"))?
+            .to_string();
+        let port = url.port().unwrap_or(DEFAULT_RTSP_PORT);
+
+        let addr = format!("{}:{}", host, port);
+        info!("Connecting to RTSP server at {}", addr);
+        let stream = TcpStream::connect(addr).await?;
+
+        Ok(RtspSession {
+            stream,
+            rtsp_url: rtsp_url.to_string(),
+            username,
+            password,
+            realm: String::new(),
+            nonce: String::new(),
+            cseq: 1,
+        })
+    }
+
+    async fn send_request(&mut self, request: &str) -> Result<()> {
+        info!("Sending RTSP request: {}", request);
+        self.stream.write_all(request.as_bytes()).await?;
+        Ok(())
+    }
+
+    async fn read_response(&mut self) -> Result<String> {
+        let mut buf = vec![0; 4096];
+        let n = self.stream.read(&mut buf).await?;
+        let response = String::from_utf8_lossy(&buf[..n]).to_string();
+        info!("Received RTSP response: {}", response);
+        Ok(response)
+    }
+
+    async fn handle_auth_response(
+        &mut self,
+        request_template: &str,
+        uri: &str,
+        method: &str,
+        cseq: u32,
+        realm: String,
+        nonce: String,
+    ) -> Result<String> {
+        let auth_header = generate_authorization_header(
+            &self.username,
+            &self.password,
+            &realm,
+            &nonce,
+            uri,
+            method,
+        );
+        let auth_request = format!(
+            "{}\r\nCSeq: {}\r\nAuthorization: {}\r\nUser-Agent: {}\r\n\r\n",
+            request_template, cseq, auth_header, USER_AGENT
+        );
+        self.send_request(&auth_request).await?;
+        self.read_response().await
+    }
+
+    async fn handle_authenticate(
+        &mut self,
+        request: &str,
+        uri: &str,
+        method: &str,
+        response: String,
+    ) -> Result<String> {
+        if response.contains("401 Unauthorized") {
```

Comment:
The `401` shouldn't use `response.contains("401 Unauthorized")`

Should use RTSP `StatusCode`, Because different rtsp server response body is different

https://docs.rs/rtsp-types/latest/rtsp_types/struct.Response.html#method.status